### PR TITLE
Adds sort to command usage help output

### PR DESCRIPTION
**Pull request by trevorlinton**
When showing the usage (if no command is available) the order of commands isn't sorted, this might be a undesirable and could be put into a config setting, but I'd like to see the tolerance to default to sorted rather than as-defined.

### CODE REVIEW

**Comment by bcoe** [comment]
@trevorlinton the nice thing about not applying sorting, is folks can control the ordering of commands by definition order.

my concern with this change is it would break a lot of downstream consumers of yargs, and would need to be a breaking change.

**Comment by trevorlinton** [comment]
can I add it as an optional config (disabled by default?) the particular issue I'm under is I don't have control with the timing when new commands are registered, so the ordering can at times be random.

**Comment by bcoe** [comment]
@trevorlinton that's an interesting requirement; I'd be open to this, I've been chatting with @cspotcode about adding the ability to configure compiler settings with a flag like `.yargsConfig` which would pass config to the `yargs-parser`; what if we allowed certain yargs settings in the non-parser layer to be configured with this too, e.g., `sort-commands` which defaults to `false`.

FYI, I think you can use `String.localeCompare()` to perform the sortinng.

**Comment by bcoe** [comment]
@trevorlinton see https://github.com/yargs/yargs/pull/1262, I'm comfortable with `sort-commands` being a configuration option that can be passed to yargs via `parserConfiguration`.

**Comment by trevorlinton** [comment]
@evocateur @bcoe I added in the `sort-commands` based on the work done in #1262 is there a common branch I can do a PR/rebase into on yargs? otherwise I'm going to be pulling in all the commits/changes from #1262 and our merges become very similar (or, alternatively, merge #1262 first and leave this be until then).

File: lib/usage.js [hunk 1]
```diff
@@ -34,7 +34,6 @@ module.exports = function usage (yargs, y18n) {
   let failureOutput = false
   self.fail = function fail (msg, err) {
     const logger = yargs._getLoggerInstance()
-
```
[the remark this answers is not in the document]
Comment (trevorlinton) [reply lib/usage.js]: fixed, sorry.

File: yargs.js [hunk 2]
```diff
@@ -1147,7 +1147,6 @@ function Yargs (processArgs, cwd, parentRequire) {
       if (err instanceof YError) usage.fail(err.message, err)
       else throw err
     }
-
```
[the remark this answers is not in the document]
Comment (trevorlinton) [reply yargs.js]: fixed

File: lib/usage.js [hunk 3]
```diff
@@ -204,7 +203,7 @@ module.exports = function usage (yargs, y18n) {
       const context = yargs.getContext()
       const parentCommands = context.commands.length ? `${context.commands.join(' ')} ` : ''
 
-      commands.forEach((command) => {
+      commands.sort((a, b) => a[0] > b[0] ? 1 : -1).forEach((command) => {
```
[the remark this answers is not in the document]
Comment (trevorlinton) [reply lib/usage.js]: updated

**Comment by bcoe** [comment]
@trevorlinton I'll get #1262 landed ASAP, and then you can rebase against `master` 👍

File: lib/usage.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (evocateur) [reply lib/usage.js]: Thanks!

**Comment by bcoe** [comment]
@trevorlinton thanks for your patience, if you don't mind rebasing I'll give this a final review 👍

one note, we'll want to make sure we document the new API method appropriately.

**Comment by trevorlinton** [comment]
@bcoe I've rebased on master, you can review the file changes, it may be prudent to squash this merge considering all of the conflicts/rebasing that occurred by merging in a unmerged PR and diverging. I've added some notes in api.md on the config option, but I'm not sure how we should document this properly, I'm hoping what's there is good enough or now.

File: docs/api.md [hunk 4]
```diff
@@ -1151,8 +1151,13 @@ for details of this object
 
 <a name="parserConfiguration"></a>.parserConfiguration(obj)
 ------------
-`parserConfiguration()` allows you to configure yargs. See [yargs-parser's configuration](https://github.com/yargs/yargs-parser#configuration) for
-settings specific to `yargs-parser`.
+`parserConfiguration()` allows you to configure yargs-parser.
```
Comment (bcoe) [inline docs/api.md]: let's make this:

"`parserConfiguration()` allows you to configure advanced yargs features:"

File: docs/api.md [hunk 5]
```diff
@@ -1151,8 +1151,13 @@ for details of this object
 
 <a name="parserConfiguration"></a>.parserConfiguration(obj)
 ------------
-`parserConfiguration()` allows you to configure yargs. See [yargs-parser's configuration](https://github.com/yargs/yargs-parser#configuration) for
-settings specific to `yargs-parser`.
+`parserConfiguration()` allows you to configure yargs-parser. 
+
+An object of options should be passed in.
```
Comment (bcoe) [inline docs/api.md]: let's change this to:

`obj` accepts the following configuration options:

File: docs/api.md (hunk 4 above)
Comment (trevorlinton) [reply docs/api.md]: done.

File: docs/api.md (hunk 5 above)
Comment (trevorlinton) [reply docs/api.md]: done.

**Comment by trevorlinton** [comment]
@bcoe I've made the changes to the docs requested.

**Comment by bcoe** [comment]
@trevorlinton 👍 thank you; I'm excited that we now have a way to configure edge-case output options for folks (just like we do with the parser).

**Comment by bcoe** [comment]
@trevorlinton please try:

`npm i yargs@next`

and let me know if it does the trick for you.

**Comment by bcoe** [comment]
this is now available in `yargs@13.1.0`.

**Pull request merged**